### ISSUE TITLE
Stair-stepping in pods going from Pending to Running

### ISSUE THREAD

**Issue by spiffxp** [kubernetes/kubernetes#14216]
We'd like to get this to a more reproducable state with e2e-generated load and some consensus on what a "correct" cluster looks like.  In the meantime, here's what we've been seeing so far.

Provision a cluster in AWS using terraform + ansible
- used https://github.com/Samsung-AG/kraken with terraform.tfvars
```
node_count = 100
aws_apiserver_type = "m4.2xlarge"
aws_master_type = "m4.2xlarge"
aws_etcd_type = "m4.2xlarge"
aws_special_node_type = "m3.xlarge"
aws_node_type = "m3.medium"
kubernetes_binaries_uri = "https://s3-us-west-2.amazonaws.com/sundry-automata/hyperkube/v1.1.0-alpha.1.1007%2Bcad5f03311afff-dirty"
```
- master is nginx + controller-manager + scheduler
- apiserver talks directly to etcd node, no local etcd proxy used
- special node is reserved for expensive services via label selectors

Setup a "pause" replication controller that creates gcr.io/google_containers/pause:go pods
- used https://github.com/Samsung-AG/kraken-services/blob/master/pause/controller.yaml

Manually scale the number of pause pods to 2000.  
```kubectl scale --replicas=10 rc pause```

Replication controller creates all pod resources in Pending by the time the `kubectl` call has returned

Total time for all pods to go to Running is ~6m.  Initial burst of >1000 pods goes from Pending to Running within the first ~2m.

Then periodic stairsteps of total number of pods in Running as the scheduler binds pods in chunks
- using prometheus to gather metrics from all kubernetes components, and etcd
- using promdash dashboard: https://github.com/Samsung-AG/kraken-services/blob/master/prometheus/build/promdash/pod-scaling.json
- no observable exhaustion of resources, just period peaks which correspond to pods moving from Pending to Running

There appears to be a ~30s interval during which the scheduler goes quiet and then kicks back to life

The thinking is... there may be some rate limiters and magic constants tuned to hold back a thundering herd.  We'd like to remove those and see if we can find where that moves the raw resource bottleneck (eg: compute, network, memory, etc).

Verified with no other services running via kubectl polling
```
$ k get pods | grep -v pause
NAME          READY     STATUS    RESTARTS   AGE
$ k get services
NAME         LABELS                                    SELECTOR   IP(S)        PORT(S)
kubernetes   component=apiserver,provider=kubernetes   <none>     10.100.0.1   443/TCP
```

Scale up to 2000 pods then poll every 10s
```
#!/bin/bash

date -u; 
time kubectl scale --replicas=2000 rc pause; 
date -u;
while true; do 
  echo $(date -u) : $(kubectl get pods --selector=k8s-app=pause | grep -ic running); 
  sleep 10;
done
date -u
```

Sample output
```
$ ./scale-up.sh
Fri Sep 18 22:05:37 UTC 2015
scaled

real	0m5.213s
user	0m0.019s
sys	0m0.009s
Fri Sep 18 22:05:42 UTC 2015
Fri Sep 18 22:05:42 UTC 2015 : 302
Fri Sep 18 22:05:53 UTC 2015 : 1117
Fri Sep 18 22:06:04 UTC 2015 : 1118
Fri Sep 18 22:06:14 UTC 2015 : 1118
Fri Sep 18 22:06:25 UTC 2015 : 1118
Fri Sep 18 22:06:35 UTC 2015 : 1118
Fri Sep 18 22:06:46 UTC 2015 : 1118
Fri Sep 18 22:06:57 UTC 2015 : 1275
Fri Sep 18 22:07:07 UTC 2015 : 1275
Fri Sep 18 22:07:18 UTC 2015 : 1275
Fri Sep 18 22:07:29 UTC 2015 : 1275
Fri Sep 18 22:07:39 UTC 2015 : 1275
Fri Sep 18 22:07:50 UTC 2015 : 1431
Fri Sep 18 22:08:01 UTC 2015 : 1431
Fri Sep 18 22:08:11 UTC 2015 : 1431
Fri Sep 18 22:08:22 UTC 2015 : 1431
Fri Sep 18 22:08:33 UTC 2015 : 1579
Fri Sep 18 22:08:43 UTC 2015 : 1588
Fri Sep 18 22:08:54 UTC 2015 : 1588
Fri Sep 18 22:09:05 UTC 2015 : 1588
Fri Sep 18 22:09:15 UTC 2015 : 1588
Fri Sep 18 22:09:26 UTC 2015 : 1738
Fri Sep 18 22:09:39 UTC 2015 : 1738
Fri Sep 18 22:09:50 UTC 2015 : 1738
Fri Sep 18 22:10:01 UTC 2015 : 1738
Fri Sep 18 22:10:11 UTC 2015 : 1777
Fri Sep 18 22:10:22 UTC 2015 : 1846
Fri Sep 18 22:10:33 UTC 2015 : 1846
Fri Sep 18 22:10:43 UTC 2015 : 1846
Fri Sep 18 22:10:54 UTC 2015 : 1846
Fri Sep 18 22:11:05 UTC 2015 : 1938
Fri Sep 18 22:11:16 UTC 2015 : 1938
Fri Sep 18 22:11:28 UTC 2015 : 1938
Fri Sep 18 22:11:39 UTC 2015 : 1938
Fri Sep 18 22:11:49 UTC 2015 : 1946
Fri Sep 18 22:12:00 UTC 2015 : 2000
```

Sample output from promdash pod-scaling dashboard during a run with `--watch-cache=true` followed by a run with `--watch-cache=false`
![watch-cache](https://cloud.githubusercontent.com/assets/49258/9973574/21e1993a-5e2a-11e5-940f-ffa4c4abddbb.png)

**Comment by lavalamp** [comment]
I'd like to get #8261 added back in, to see if it makes a difference.

To nail down this further, I'd like to see scheduler logs with these three things:
* time pod enters queue
* time bind call is initiated
* time bind call completes

At least some of these things should already be logged.

**Comment by spiffxp** [comment]
Anecdotally, this appears to be a regression in the 1.1.x codebase.  

Using the exact same set of command line flags, I've been unable to reproduce stair-stepping behavior with server versions:
- v1.0.3 release
- v1.0.6 release
- v1.0.6 built locally

I've reproduced it with server versions:
- v1.1.0-alpha.1.1007+cad5f03311afff-dirty built by ci
- v1.1.0-alpha.1.1007+cad5f03311afff-dirty built locally
- (1.1.x builds locally going back at least as far as 8/5 but I don't have a specific list)

Will look at patches for #8261 and more logging next

**Comment by spiffxp** [comment]
Two runs
- patch for #8261 applied (
https://github.com/kubernetes/kubernetes/compare/master...spiffxp:reapply-8261)
- v1.0.6 release

Raised two settings:
- apiserver max-requests-inflight x100 (from 400 to 40000)
- kubelet max-pods x5 (from 40 to 200)

The patch smooths out stair steps, but doesn't really reduce overall duration.  The release has no stair steps.

<img width="958" alt="reapply-8261-vs-v1 0 6" src="https://cloud.githubusercontent.com/assets/49258/10009332/d03cef72-608f-11e5-8b12-d6ad67cb60d5.png">

**Comment by davidopp** [comment]
@kubernetes/sig-scalability

**Comment by wojtek-t** [comment]
@spiffxp - did you managed to confirm that the scheduler is guilty for stair-stepping? Maybe it's Kubelet? Or watch propagation? Or sth else (like etcd?)

**Comment by lavalamp** [comment]
Thoughts:
* kubelet_docker_ops_latency_seconds shows a noticeable difference between the two runs. Is the same version of docker running in both configurations? 'inspect container' and 'list containers' specifically seem to take a lot more time. @dchen1107 any thoughts on this?
* patch for #8261 smooths out stairsteps without fixing the problem, indicating that the stairsteps are at best a symptom, not a cause.

**Comment by timothysc** [comment]
@wojtek-t we've pretty much eliminated etcd as a culprit.

**Comment by spiffxp** [comment]
Ran density 30 pod test case against the cluster configuration described above.  The only difference between each run is the hyperkube binary used.  

Unfortunately the latencies don't really reflect the difference in performance, was hoping this could be tied / related to #14554
```
v1.0.6

Sep 25 15:38:29.209: INFO: 10% worst schedule latencies: 
Sep 25 15:38:29.209: INFO: perc50: 0, perc90: 0, perc99: 0
Sep 25 15:38:29.209: INFO: 10% worst run-after-schedule latencies: 
Sep 25 15:38:29.209: INFO: perc50: 3s, perc90: 4s, perc99: 7s
Sep 25 15:38:29.209: INFO: 10% worst watch latencies: 
Sep 25 15:38:29.209: INFO: perc50: 2m28.898824401s, perc90: 2m29.344273638s, perc99: 2m30.224033392s
Sep 25 15:38:29.209: INFO: 10% worst scheduled-to-end total latencies: 
Sep 25 15:38:29.210: INFO: perc50: 2m31.819693166s, perc90: 2m33.17225807s, perc99: 2m36.069894414s
Sep 25 15:38:29.210: INFO: 10% worst e2e total latencies: 
Sep 25 15:38:29.210: INFO: perc50: 2m31.819693166s, perc90: 2m33.17225807s, perc99: 2m36.069894414s

0662141

Sep 25 16:31:06.037: INFO: 10% worst schedule latencies: 
Sep 25 16:31:06.037: INFO: perc50: 0, perc90: 0, perc99: 0
Sep 25 16:31:06.037: INFO: 10% worst run-after-schedule latencies: 
Sep 25 16:31:06.037: INFO: perc50: 3s, perc90: 7s, perc99: 9s
Sep 25 16:31:06.037: INFO: 10% worst watch latencies: 
Sep 25 16:31:06.037: INFO: perc50: 2m29.948415372s, perc90: 2m43.400006051s, perc99: 2m54.000208969s
Sep 25 16:31:06.037: INFO: 10% worst scheduled-to-end total latencies: 
Sep 25 16:31:06.037: INFO: perc50: 2m34.405696183s, perc90: 2m45.400006051s, perc99: 2m55.401601692s
Sep 25 16:31:06.037: INFO: 10% worst e2e total latencies: 
Sep 25 16:31:06.038: INFO: perc50: 2m34.405696183s, perc90: 2m45.400006051s, perc99: 2m55.401601692s
```

v1.0.6
<img width="1911" alt="100-m3medium-kube-v106-density-30" src="https://cloud.githubusercontent.com/assets/49258/10114314/ead152a8-63a3-11e5-86b1-c49e366420bd.png">

0662141
<img width="1913" alt="100-m3medium-kube-master-density-30" src="https://cloud.githubusercontent.com/assets/49258/10114316/f5305ac8-63a3-11e5-9c27-22b02fb8062a.png">

push gateway metrics from 0662141
<img width="1861" alt="100-m3medium-kube-master-density-30-e2e" src="https://cloud.githubusercontent.com/assets/49258/10114326/2071959e-63a4-11e5-818b-afca2d7d52b2.png">

**Comment by spiffxp** [comment]
Added a resource block to https://github.com/Samsung-AG/kraken-services/blob/master/pause/controller.yaml
```
resources:
  requests:
    memory: "16Mi"
    cpu: "25m"
```

Manually scaled to 3030 replicas for parity with density test, immediate stair stepping

<img width="1910" alt="100-m3medium-kube-master-scale-resource-request" src="https://cloud.githubusercontent.com/assets/49258/10114560/89000d7c-63a7-11e5-9a52-91e078c3f020.png">

**Comment by spiffxp** [comment]
These are the kubernetes flags / constants used during these runs https://github.com/Samsung-AG/kraken/blob/master/ansible/roles/kubernetes/defaults/main.yaml#L10-L78

Most of them should be the defaults, just made explicit.  Tweaks have the default commented out above.

**Comment by dchen1107** [comment]
cc/

**Comment by wojtek-t** [comment]
>
These are the kubernetes flags / constants used during these runs https://github.com/Samsung-AG/kraken/blob/master/ansible/roles/kubernetes/defaults/main.yaml#L10-L78

@spiffxp - did you run exactly the same test with all default values? In particular I'm wondering whether this line:
"max-requests-inflight: 40000"
is causing some problems here.

**Comment by spiffxp** [comment]
Yes, same flags, same test case used on both runs.  The only difference was the version of hyperkube used.  

Anecdotally I've tweaked the max-requests-inflight value from 400-4000000 with no visible effect, but I will see if I can get another run with all defaults today.

For me the more interesting thing is that adding resource constraints triggered stair-stepping almost immediately.

**Comment by wojtek-t** [comment]
>
For me the more interesting thing is that adding resource constraints triggered stair-stepping almost immediately.

That's pretty interesting - FYI @kubernetes/goog-control-plane

**Comment by gmarek** [comment]
@spiffxp - this is indeed very interesting. Were you running your pods inside a default namespace, or created your own (and if you did, did you have a LimitRanger for it) - it's important to know if it's introducing resources at all, or changing default values that had this effect. Second thing is - have you measured schedule latency? AFAICT resource requests/limits should impact only a scheduler, and in density tests we see that scheduling is pretty much instantaneous.

I'm going to change our Density test to set some requests and see what impact it'll have.

**Comment by gmarek** [comment]
The results with requests introduced are the same as without them on my end.

**Comment by spiffxp** [comment]
Adding resource requests / limits triggers stair-stepping on v1.0.6 almost immediately as well, so I can't quite git-bisect this one away like I'd hoped.  

@gmarek these pods are in the default namespace, all I recall doing is using the recommended list of admission controllers: https://github.com/Samsung-AG/kraken/blob/master/ansible/roles/kubernetes/defaults/main.yaml#L12

I'll get scheduler binding / algorithm / e2e latency in the next set of graphs

**Comment by gmarek** [comment]
I'm sure - we do have LimitRange in v1.0.6, right? @bgrant0607 

If so, then something strange is happening, as they would have a default Limits inherited from `default` namespace LimitRanger...

**Comment by lavalamp** [comment]
The e2e test makes its own namespace, so these test pods wouldn't be in the default namespace.

**Comment by lavalamp** [comment]
@spiffxp do the resource limits increase the overall time in 1.0.6?

**Comment by lavalamp** [comment]
This is very interesting. Scheduler behaves differently (by design) when there are resources assigned, and so do kubelets. So that narrows it down to one of those two components. It could be something like scheduler is going at the same rate, but making worse choices.

If kubelet is involved, I'd expect to see some phase=Failed pods as kubelet rejects over-committed pods.

**Comment by bgrant0607** [comment]
LimitRange exists in 1.0.6, yes. In our standard GCE setup, the LimitRanger admission control plugin is enabled and we create a LimitRange for the default namespace, which set cpu to 100m for all containers by default.

**Comment by gmarek** [comment]
@lavalamp - AFAIU @spiffxp don't run e2e test, but start containers by hand in a similar number. The thing is, that if they are put into default namespace, they always have limits set, which would mean that *changing* limits makes the difference. Or I misunderstood something completely.

**Comment by bgrant0607** [comment]
@gmarek Limits, requests, or both?

**Comment by gmarek** [comment]
I don't know - @spiffxp? We don't have requests working in 1.0.6, right?

**Comment by spiffxp** [comment]
@gmarek I think you're right, I don't think resource requests were introduced into the release-1.0 branch.  

Using only resource requests does not trigger stair stepping on v1.0.6

Using only resource limits triggers stair stepping on v1.0.6

Using either/both of resource requests/limits triggers stair stepping on release-1.1

`kubectl limits` shows no limits in place for any namespace, does that mean the default LimitRange doesn't exist, or do we just not see it?

**Comment by spiffxp** [comment]
Moving from single apiserver to three apiservers had no effect

Setting all flags to default values except kubelet's server flag (max-requests-inflight, max-pods, cadvisor-port) had no effect

**Comment by lavalamp** [comment]
Given that *request* doesn't trigger the problem but *limit* does, I think
we should focus on kubelet. The scheduler cares only about request:
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go

Do you get rejected pods? Try running the below command during/after a run:

kubectl get pods -o template --template="{{range .items}}{{if eq
.status.phase \"Failed\"}}{{.metadata.name}}"$'\n'"{{end}}{{end}}" | wc -l

**Comment by spiffxp** [comment]
@lavalamp resource limits increase the overall time in v1.0.6 and increase the `scheduler_scheduling_algorith_latency_microseconds_sum` metric

First a `kubectl scale --replicas=300 rc pause` with resource limits in place, eg:
```yaml
spec:
  containers:
    - name: pause
      image: gcr.io/google_containers/pause:go
      resources:
        limits:
          memory: "16Mi"
          cpu: "100m"

```
Then a `kubectl scale --replicas=300 rc pause` with no resource limits in place, eg:
```yaml
spec:
  containers:
    - name: pause
      image: gcr.io/google_containers/pause:go
```

![scheduling-algorithm-latency](https://cloud.githubusercontent.com/assets/49258/10323670/6826fff8-6c39-11e5-94fc-611d4512783d.png)

**Comment by lavalamp** [comment]
Also, what is your workload? Does it spike resource usage on startup? If
so, kubelet limiting it could legitimately make startup much slower!

**Comment by spiffxp** [comment]
Not sure I quite understand the workload question, I'm assuming you mean what other pods/services are we running?  By default we run with a few services which I could see as generating kubelet overhead (https://github.com/Samsung-AG/kraken/blob/master/terraform/aws/variables.tf#L238, corresponds to directories in https://github.com/Samsung-AG/kraken-services)

But as I said up top, I've repro'ed this with no other services running, verifying via a kubectl polling loop

**Comment by lavalamp** [comment]
Sorry, I see you're running the pause pod-- that's all I was asking. :)

I'm reversing my previous guess after an IRL talk here. I now predict you'll see no failed pods if you run the command I gave a couple comments ago.

**Comment by lavalamp** [comment]
I've made a scheduler integration benchmark test.

First experiment: how does per-pod time vary with number of nodes?

10 nodes: 68ms
100 nodes: 74ms
1000 nodes: 90ms

That's not even linear, great. I had to disable the client rate limiting.

**Comment by lavalamp** [comment]
@spiffxp can you post the exact command line you're running kube-scheduler with?

**Comment by lavalamp** [comment]
Final test of the day: removing the resources from the pod reduces the 90ms number above down to 74ms. (my test doesn't turn on all the fancy defaulting things in 1.1/head.)

My hypothesis at the moment is that 1.1 is just much more effective at defaulting the pod's resource requests, which triggers a slow case in the scheduler. I'm not sure what the slow case is exactly, but the next thing I'll check is whether the scheduler is making extra API calls to in this case, consuming its QPS quota.

**Comment by spiffxp** [comment]
@lavalamp I don't have the command handy but the flags come from https://github.com/Samsung-AG/kraken/blob/master/ansible/roles/kubernetes/defaults/main.yaml#L75-L79 and https://github.com/Samsung-AG/kraken/blob/master/ansible/roles/kubernetes/templates/hyperkube-binary-scheduler.service.ansible#L14-L15.  The `master_private_ip` variable corresponds to an nginx instance which (potentially) fronts multiple apiserver instances, but I'm pretty sure I was initially seeing this on a cluster with api+sched+cm colocated.  Maybe worth a revisit.

It's been a few days since I last played with raising api limits, can PR https://github.com/kubernetes/kubernetes/commit/124b6d88d1646af77ff98ea2e04134d7c907b268 if it's worthwhile

**Comment by lavalamp** [comment]
@spiffxp When you ran 124b6d8 (which may be worth a PR, btw), what values did you set for the flags?

**Comment by gmarek** [comment]
@lavalamp: I don't think it's scheduler, as even 90 ms is good enough (AFAICT scheduler does 2 API server calls per scheduling, which with 20 QPS limit gives ~100ms latency - for some meaning of it). Moreover in results pasted by @spiffxp here https://github.com/kubernetes/kubernetes/issues/14216#issuecomment-143378124 it doesn't look like we have any problems with the scheduler - it's the Kubelet that it's slow (run after schedule, i.e. difference between the time when all Containers are running [understood as a latest StartedAt timestamp] and the time Scheduler emitted a 'scheduled' event, is surprisingly large).

There are few things that are happening during this step:
- scheduler writes Binding to etcd
- Kubelet's watch gets created binding
- Kubelet processes request, starts all containers and creates a new PodStatus object (no writing here, as we're looking at timestamp created by Kubelet).

Out of curiosity - @spiffxp are clocks in your cluster synchronized?

When running your experiment on a fresh cluster can you get the result of `curl -L http://127.0.0.1:4001/v2/stats/store` (or equivalent) from etcd machine just after you finish? Maybe something is writing a lot more stuff to etcd...

Other ideas: you can try looking at Kubelet metrics (`/metrics` endpoint on Kubelet's read-only port (10255 by default)) and see if there's nothing surprisingly large there (interesting stuff is close to the end).

**Comment by lavalamp** [comment]
@gmarek I don't think we can rule out the scheduler. If you look at the chart in https://github.com/kubernetes/kubernetes/issues/14216#issuecomment-146014551, @spiffxp is clearly seeing much higher latency numbers than my test produces.

**Comment by spiffxp** [comment]
@gmarek I don't think we explicitly synchronize our clocks, will look into that today

I agree with @lavalamp about not yet ruling out scheduler, the density latencies I posted were using the default version of the test with no resource limits in place.  I expect getting numbers from https://github.com/spiffxp/kubernetes/commit/65ea246cefd729bcf055a618b175f3a6d4432dfe will look different

Will also double-check if raising kube-api limits via #15247 helps

**Comment by lavalamp** [comment]
I played with my test a bit more today, but wasn't able to conclusively reproduce anything.

**Comment by wojtek-t** [comment]
>
I played with my test a bit more today, but wasn't able to conclusively reproduce anything.

I was also playing with 100-node clusters and I just cannot reproduce it.

**Comment by gmarek** [comment]
I was playing with 1000 node simulated cluster and noticed, that scheduler indeed isn't saturated in the beginning. It uses roughly 80% of CPU for first few minutes, which suggests that there may be some contention there (it starts using whole CPU as the load grows, but it takes few minutes).

**Comment by HaiyangDING** [comment]
CC @HaiyangDING

**Comment by wojtek-t** [comment]
@gmarek - I think that it's because initially, "scheduling" is relatively faster than "submitting binding to apiserver and waiting for the response" (which basically consumes no cpu). Other reason might be that we are scheduling faster than the pods for scheduling are coming.

**Comment by gmarek** [comment]
Second thing is not the case - there already was a queue of 'waiting' pods; and first is a kind of contention.

**Comment by spiffxp** [comment]
Clocks are synchronized via the default mechanisms in place on CoreOS

<img width="959" alt="api-rates" src="https://cloud.githubusercontent.com/assets/49258/10461586/f8dd9674-71a7-11e5-9803-b35648792a9a.png">

Cluster configuration same as the top of this issue, but 40 nodes instead of 100

density-3 w/ cpu and mem limits in the RC

Left to right
1. kube-api-burst/qps, bind-pods-burst/qps / 10 for all components except controller-manager (left at defaults)
2. max-requests-inflight set to 1
3. scheduler rates * 2
4. scheduler rates * 2
5. scheduler rates * 2 (stair stepping appears)
6. **control** all rates at defaults
7. kube-api-burst/qps, bind-pods-burst/qps, max-requests-inflight * 100

Observations
- all runs reached their expected pod count, the peaks look lower because prometheus polling interval may have missed max pods
- max-requests-inflight appears to have no effect?  I was hoping to cause a thundering herd just to know what it looks like
- kubelet/proxy api rate limits had no observable effects so I left them out of this

I'm wondering if we're running into this magic constant https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/modeler.go#L108

**Comment by spiffxp** [comment]
Nope, I raised that constant to 300s and down to 1s with no visible effect on behavior.

**Comment by lavalamp** [comment]
Logs please!

**Comment by spiffxp** [comment]
https://s3-us-west-2.amazonaws.com/k8s-issue-14216/logs.tgz
- all command line flags at defaults, except scheduler's `--v=4`
- bad: 2 runs with resource limits in place
- good: 2 runs without resource limits in place
```
$ tar -tf logs.tgz
logs/
logs/bad.controller-manager.log
logs/bad.density-runs.log
logs/bad.scheduler.log
logs/good.controller-manager.log
logs/good.density-runs.log
logs/good.scheduler.log
```

**Comment by lavalamp** [comment]
@spiffxp I will take a look, but apiserver logs are probably what I want to see, can you include them?

**Comment by lavalamp** [comment]
@spiffxp Can you pull #15890? I haven't actually looked at anything that I intended to look at yet, but this is a glaring difference.

**Comment by lavalamp** [comment]
I will be very sad if --v=4 on scheduler turns out to have been the relevant difference :(

**Comment by spiffxp** [comment]
As a quick sanity check I ran with --v=0 for everything, and yup, stair steps went away on the smaller cluster.  Cue facepalm.

Will merge in the PR and try against larger cluster configurations later today.

**Comment by gmarek** [comment]
FYI: we're running with --v=2

**Comment by lavalamp** [comment]
I guess this explains why I couldn't get latency numbers anything like what you reported in my benchmark!

**Comment by lavalamp** [comment]
@timothysc Does the above explain the stairstepping you previously reported?

**Comment by spiffxp** [comment]
OK, I've verified that running with the original 100 node configuration using v1.2.x, v1.1.x, v1.0.6 with the scheduler at `--v=1` or v1.2.x with #15890 applied with the scheduler at `--v=4` looks identical for the density-3 and density-30 cases.

@lavalamp though not in an N^2 path, there are still some pretty verbose logs in the scheduler loop if `namespace/pod` is sufficient to trace the flow of scheduling, eg: [this attempting to schedule message at `--v=3`](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/scheduler.go#L119)

Willing to call this closed if we're sufficiently satisfied we've found root cause

**Comment by lavalamp** [comment]
@spiffxp I think that one is OK, you see it once per scheduling attempt and it has some useful info. Let's see if we can hear from @timothysc before we close this.

It's also worth thinking about why this caused stair stepping instead of uniform slowness. Was it the OS batching log file writes? Something else? On GCE I'd guess I/O throttling, but you shouldn't have that on bare metal.

**Comment by spiffxp** [comment]
Based on discussion in the sig-scale hangout last week, sounds like we're comfortable closing this as-is for now.  If @timothysc ends up re-encountering, we can always reopen.

**Comment by spiffxp** [comment]
I'm seeing this behavior again in v1.1.1.  At this point I'm pretty sure something is wrong with our journald setup.  In the meantime, still trying to keep an eye on excessively verbose scheduler messages.  #17004 will take care of some of this, and hopefully #17348 catches the rest (for now.)  At that point I can run a scheduler with --v=1 and not run directly into whatever our journald issue is.

**Comment by spiffxp** [comment]
Workaround: bypass `journald` entirely.  Either send StandardError to null, or set `--logtostderr=false` and have the kubernetes components log directly to files on their own.

https://github.com/kubernetes/kubernetes/blob/master/pkg/util/logs.go#L32

Based on the TODO here I'm thinking we should revisit whether `--logtostderr=true` is an appropriate default.  It appears most of the saltbase stuff redirects stderr/stdout to a file rather than relying on the kubernetes components to do the logging themselves.

**Comment by spiffxp** [comment]
#17438 has been merged into master and cherry-picked into release-1.1, happy to call this closed once a new 1.1.x release has been cut
